h/static/scripts/websocket.js

```diff
@@ -32,44 +38,80 @@ function Socket(url) {
     }
   }
 
-  function reconnect() {
-    var didConnect = false;
-    var connectOperation = retry.operation();
-    connectOperation.attempt(function (currentAttempt) {
+  // Connect the websocket immediately. If a connection attempt is already in
+  // progress, do nothing.
+  function connect() {
+    if (operation) {
+      return;
+    }
+
+    operation = retry.operation({
+      minTimeout: RECONNECT_MIN_DELAY * 2,
+      // Don't retry forever -- fail permanently after 10 retries
+      retries: 10,
+      // Randomize retry times to minimise the thundering herd effect
+      randomize: true
+    });
+
+    operation.attempt(function () {
       socket = new WebSocket(url);
       socket.onopen = function (event) {
-        // signal successful connection
-        connectOperation.retry();
-        didConnect = true;
-        sendMessages();
-
+        onOpen();
         self.emit('open', event);
       };
-
       socket.onclose = function (event) {
-        if (event.code !== CLOSE_NORMAL) {
-          if (didConnect) {
-            console.warn('The WebSocket connection closed abnormally ' +
-              '(code: %d, reason: %s). Reconnecting automatically.',
-              event.code, event.reason);
-            reconnect();
-          } else {
-            console.warn('Retrying connection (attempt %d)', currentAttempt);
-            connectOperation.retry(new Error(event.reason));
-          }
+        if (event.code === CLOSE_NORMAL) {
+          self.emit('close', event);
+          return;
         }
-        self.emit('close', event);
+        var err = new Error('WebSocket closed abnormally, code: ' + event.code);
+        console.warn(err);
+        onAbnormalClose(err);
       };
-
       socket.onerror = function (event) {
         self.emit('error', event);
       };
-
       socket.onmessage = function (event) {
         self.emit('message', event);
       };
     });
-  };
+  }
+
+  // Reconnect the websocket after a short delay. This is used when the
+  // websocket has closed abnormally. If reconnection is already in progress, do
+  // nothing.
+  function reconnect() {
+    if (operation) {
+      return;
+    }
+    var delay = RECONNECT_MIN_DELAY;
+    delay += Math.floor(Math.random() * delay);
+    operation = setTimeout(function () {
+      operation = null;
+      connect();
+    }, delay);
+  }
+
+  // onOpen is called when a websocket connection is successfully established.
+  function onOpen() {
+    operation = null;
+    sendMessages();
+  }
+
+  // onAbnormalClose is called when a websocket connection closes abnormally.
+  // This may be the result of a failure to connect, or an abnormal close after
+  // a previous successful connection.
+  function onAbnormalClose(error) {
+    // If we're already in a reconnection loop, trigger a retry...
+    if (operation) {
+      if (!operation.retry(error)) {
+        console.error('reached max retries attempting to reconnect websocket');
```

Comment:
Hmm. We're not clearing `operation` here, even though an operation is no longer in progress after this. Strictly speaking it doesn't matter though because `Socket` doesn't provide a way to initiate a reconnection externally.